### DISCUSSION TITLE
@planetscale/database - A couple of questions about creating connections

### DISCUSSION THREAD

**Discussion by cdcarson** [planetscale/discussion#383] [Q&A] [answered]
Background: I'm currently using [Cloudflare Pages](https://developers.cloudflare.com/pages/) to build an app. I have very little idea about when Cloudflare isolates are created/destroyed , but [it seems like a single isolate handles multiple requests and could be long lived](https://developers.cloudflare.com/workers/learning/how-workers-works/#distributed-execution).

I'm currently creating a singleton PlanetScale connection for each instance of my app -- i.e., each isolate would share the connection across multiple user requests and the PlanetScale connection would live as long as the isolate. I _was_ pretty sure this would be OK. The PlanetScale connection seemed like just a config object that didn't maintain an actual network connection or even make one until used for a query or transaction. 

But then I saw that `Connection` has a `refresh()` method. So question 1:

- What does `connection.refresh()` do, and do I need to worry about it?

Related question, more general:

- Is there a best practice for creating connections? Is there a downside to having lots of them, like on a per web request or even on a per query/transaction basis? Conversely, is there a downside to keeping one around a long time (as I'm doing now?)

Thanks in advance for any help.

**Comment by orware** [comment, marked as answer]
Hello @cdcarson,

I may loop in some additional folks from our end to provide some extra insight for you on your questions, but based on what you've shared so far I wouldn't expect you to run into issues. 

The new [PlanetScale Global Routing Infrastructure](https://planetscale.com/blog/one-million-connections#one-million-mysql-connections) is able to handle lots of connections, but your workload ultimately will be the limiting factor depending on how the queries being sent to it are processed along with the resulting load that is created on your underlying MySQL instance. 

Keeping a connection around for an extended period as well should generally not be problematic, and in most cases may be preferred mainly because that helps to improve responsiveness due to the connection already being established so long as your application is able to re-establish connections automatically.

The `refresh()` method you mentioned mainly runs `createSession()` [under the hood](https://github.com/planetscale/database-js/blob/main/src/index.ts#L192). 

To my understanding, the HTTP/2 (or HTTP/3) based session created here with our serverless JavaScript driver in this case acts as a stand-in for a normal connection you might create within an application that is connecting via the MySQL protocol.

Within [database-js](https://github.com/planetscale/database-js) the connection/session is created automatically "just in time" when you issue the first official query, which will generally add a bit of delay for the first query's response to be returned as the session gets established for the first time. Subsequent queries would then be faster since the session is already available to be reused.

You can potentially make sure this is already done before running any queries by running `refresh()` manually to establish the session sooner, but generally that shouldn't be required.

**Reply by mattrobenolt** [reply]
Yeah, to reiterate some of the bits, an HTTP connection (which is what comes from database-js) is pretty close to free to us, much free-er than a MySQL connection. You also can multiplex many sessions across one single HTTP connection, since the connections are stateless.

So in practice, you only need one real HTTP connection to us, with multiple sessions. In this case, multiple configs since you also need multiple different credentials.

Credentials are passed per-request and the session is handled similar to a typical browser cookie. So per-request, the session and credentials are passed thus making the entire connection stateless.

As long as the credentials and session state don't get mixed up, this is entirely safe to do everything over one single HTTP request, unlike MySQL where you need to manage a connection pool for a bunch of connections.

So I'd say to directly answer your questions, best practice would be to have as few HTTP connections just to minimize the overhead of creating new ones, and everything can safely be shared over that single connection. Whether that's the entire lifetime of an application or lifetime of a single request, etc, it doesn't particularly matter to us. You'd gain more benefits by just reusing a single connection as long as possible, and sharing it across many different credentials and sessions.

I hope that helps shed some light.

**Reply by cdcarson** [reply]
Hey @orware and @mattrobenolt, thanks for the detailed and helpful replies.  This is a new paradigm for me.

I'll continue doing what I'm doing now, one connection in the scope of an isolate.

**Comment by Izhaki** [comment]
@orware 

> Within [database-js](https://github.com/planetscale/database-js) the connection/session is created automatically "just in time" when you issue the first official query,

This seems to answer a similar question I have, but just want to verify.

Basically, for local runs, do we need something similar to [this](https://github.com/prisma/prisma-examples/blob/36eccfac9e7feae1ae2b2758d2066f7bce0430c6/typescript/rest-nextjs-api-routes-auth/lib/prisma.ts)?

```ts
// PrismaClient is attached to the `global` object in development to prevent
// exhausting your database connection limit.
//
// Learn more:
// https://pris.ly/d/help/next-js-best-practices
const globalForPrisma = global as unknown as { prisma: PrismaClient };

export const prisma =
  globalForPrisma.prisma ||
  new PrismaClient();

if (process.env.NODE_ENV !== 'production') globalForPrisma.prisma = prisma;
```

The [f1-championship-stats](https://github.com/planetscale/f1-championship-stats) example goes:

```js
const db = new Client({
  url: process.env['DATABASE_URL']
})

export default async function handler(req: Request) {
  const conn = db.connection()
  // ...
  conn.execute(...);
}
```

Will this create multiple connections on local?

What about this?

```js
const db = new Client({
  url: process.env['DATABASE_URL']
})

const conn = db.connection()

export default async function handler(req: Request) {
  conn.execute(...);
}
```

Thanks!
